vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
+
+    if (attr_obj_id == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_OBJ_ID was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    sai_object_id_t obj_id = attr_obj_id->value.oid;
+
+    sai_object_type_t ot = sai_object_type_query(obj_id);
+
+    if (ot != SAI_OBJECT_TYPE_PORT)
+    {
+        SWSS_LOG_ERROR("SAI_HOSTIF_ATTR_OBJ_ID=%s expected to be PORT but is: %s",
+                sai_serialize_object_id(obj_id).c_str(),
+                sai_serialize_object_type(ot).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // validate SAI_HOSTIF_ATTR_NAME
+
+    auto attr_name = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_NAME, attr_count, attr_list);
+
+    if (attr_name == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_NAME was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strnlen(attr_name->value.chardata, sizeof(attr_name->value.chardata)) >= MAX_INTERFACE_NAME_LEN)
+    {
+        SWSS_LOG_ERROR("interface name is too long: %.*s", MAX_INTERFACE_NAME_LEN, attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strncmp(attr_name->value.chardata, "Ethernet", 8) != 0)
+    {
+        SWSS_LOG_ERROR("interface name should start with EthernetX but is %s", attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    std::string name = std::string(attr_name->value.chardata);
+
+    // create TAP device
+
+    SWSS_LOG_INFO("creating hostif %s", name.c_str());
+
+    int tapfd = vs_create_tap_device(name.c_str(), IFF_TAP | IFF_MULTI_QUEUE | IFF_NO_PI);
+
+    if (tapfd < 0)
+    {
+        SWSS_LOG_ERROR("failed to create TAP device for %s", name.c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    SWSS_LOG_INFO("created TAP device for %s, fd: %d", name.c_str(), tapfd);
+
+    // XXX currently tapfd is ignored, it should be closed on hostif_remove
+    // and we should use it to read/write packets from that interface
+
+    sai_attribute_t attr;
+
+    attr.id = SAI_SWITCH_ATTR_SRC_MAC_ADDRESS;
+
+    sai_status_t status = vs_generic_get(SAI_OBJECT_TYPE_SWITCH, switch_id, 1, &attr);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("failed to get SAI_SWITCH_ATTR_SRC_MAC_ADDRESS on switch %s: %s",
+                sai_serialize_object_id(switch_id).c_str(),
+                sai_serialize_status(status).c_str());
+    }
+
+    int err = vs_set_dev_mac_address(name.c_str(), attr.value.mac);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("failed to set MAC address %s for %s",
+                sai_serialize_mac(attr.value.mac).c_str(),
+                name.c_str());
+
+        close(tapfd);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    err = ifup(name.c_str());
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("failed to bring ifup %s", name.c_str());
+
+        close(tapfd);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (!hostif_create_tap_veth_forwarding(name, tapfd))
```

Comment:
I think it is better to open the peer link netdev in the VS init functions. We can then assign each port a lane information, and return the lane information to the orchagent and a port id. In the create_hostif, we just attach the host interface (tap device) to the peer link based on the port id.

This work flow is more like the SAI flow.

Reply:
peer link is assigned when hostif is created, at VS init its not know yet, currenty all functions are asynchronous and orchagent has no idea about any return information, also lane information also can be obtained and assigned after hostif create using GET sai functions

Reply:
orchagent will query the lane information first, is that also asynchronous even for the get API?

Reply:
get is synchronous, but why you want to create hostif at vs init ?

Reply:
I do not mean create hostif at the vs init. I mean just open the virtual link at the vs init. The virtual link is the physical interface of the virtual switch. They should exist whether or not there is host interface. For example, you can get link status, notification, learn mac regardless of host interface creation or not.

I am thinking, maybe we should open the vpeer link at the vs init and start processing packet, maybe doing the mac learning and a few other things. Only when when create hostif then we can transferring the packets between the vpeer link and the tap devices.

For now, I am ok.